README.rst
===================================================
Community Detection gProfiler Gene To Term Mapper
===================================================

This repository creates a CDAPS compatible functional enrichment Docker image using
`gProfiler <https://pypi.org/project/gprofiler-official/>`_

Dependencies
------------

* `Docker <https://www.docker.com/>`_
* `make <https://www.gnu.org/software/make/>`_ (to build)
* Python (to build)

Direct invocation
------------------

Version `0.3.0` can be directly pulled from `Dockerhub <https://hub.docker.com/>`_ with this command:

.. code-block::

   docker pull coleslawndex/cdgprofilergenestoterm:0.3.0

Building
--------

.. code-block::

   git clone https://github.com/idekerlab/cdgprofilergenestoterm
   cd cdgprofilergenestoterm
   make dockerbuild

Run **make** command with no arguments to see other build/deploy options including creation of Docker image

.. code-block::

   make

Output:

.. code-block::

   clean                remove all build, test, coverage and Python artifacts
   clean-build          remove build artifacts
   clean-pyc            remove Python file artifacts
   clean-test           remove test and coverage artifacts
   lint                 check style with flake8
   test                 run tests quickly with the default Python
   test-all             run tests on every Python version with tox
   coverage             check code coverage quickly with the default Python
   docs                 generate Sphinx HTML documentation, including API docs
   servedocs            compile the docs watching for changes
   testrelease          package and upload a TEST release
   release              package and upload a release
   dist                 builds source and wheel package
   install              install the package to the active Python's site-packages
   dockerbuild          build docker image and store in local repository
   dockerpush           push image to dockerhub


Usage
-----

.. code-block::

   docker run -v coleslawndex/cdgprofilergenestoterm:0.3.0 -h




Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.3.0 (2020-02-24)
---------------------

* Added *--minoverlap* which lets user filter results by Jaccard

* Added *--excludesource* which lets user exclude sources from result

* Changed output to include **sourceTermId** and **jaccard**

0.2.0 (2019-12-12)
---------------------

* Default pvalue set to 1e-8

* Added term_size to json output which contains number of genes associated with term


0.1.0 (2019-09-19)
------------------

* First release on PyPI.


tests/test_cdgprofilergenestoterm.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_cdgprofilergenestoterm
----------------------------------

Tests for `cdgprofilergenestoterm` module.
"""

import os
import sys
import unittest
import tempfile
import shutil
from unittest.mock import MagicMock
import pandas as pd

from cdgprofilergenestoterm import cdgprofilergenestotermcmd


class TestCdgprofilergenestoterm(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_read_inputfile(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('hellothere')
            res = cdgprofilergenestotermcmd.read_inputfile(tfile)
            self.assertEqual('hellothere', res)
        finally:
            shutil.rmtree(temp_dir)

    def test_parse_args(self):
        myargs = ['inputarg']
        res = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                         myargs)
        self.assertEqual('inputarg', res.input)
        self.assertEqual(0.00000001, res.maxpval)
        self.assertEqual(0.05, res.minoverlap)
        self.assertEqual('hsapiens', res.organism)
        self.assertEqual(500, res.maxgenelistsize)

    def test_run_gprofiler_no_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)
            try:
                cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                        theargs)
                self.fail('Expected FileNotFoundError')
            except FileNotFoundError:
                pass
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)
            res = cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                          theargs)
            self.assertEqual(None, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_with_empty_result(self):
        temp_dir = tempfile.mkdtemp()
        try:
            mygprofiler = MagicMock()
            mygprofiler.profile = MagicMock(return_value=pd.DataFrame())
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c')
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)
            res = cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                          theargs,
                                                          gprofwrapper
                                                          =mygprofiler)
            self.assertEqual(None, res)
            mygprofiler.profile.assert_called_once_with(query=['a', 'b', 'c'],
                                                        domain_scope='known',
                                                        organism='hsapiens',
                                                        user_threshold=0.00000001,
                                                        no_evidences=False)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_with_too_many_genes(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')

            # write out gene file with 5002 genes which
            # should return None
            with open(tfile, 'w') as f:
                for x in range(0, 5001):
                    f.write(str(x) + ',')
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)
            res = cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                          theargs)
            self.assertEqual(None, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_with_empty_result(self):
        temp_dir = tempfile.mkdtemp()
        try:
            mygprofiler = MagicMock()

            df = pd.DataFrame()
            mygprofiler.profile = MagicMock(return_value=df)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c,')
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)
            res = cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                          theargs,
                                                          gprofwrapper=
                                                          mygprofiler)
            self.assertEqual(None, res)

            mygprofiler.profile.assert_called_once_with(query=['a', 'b', 'c'],
                                                        domain_scope='known',
                                                        organism='hsapiens',
                                                        user_threshold=0.00000001,
                                                        no_evidences=False)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_with_no_result_meeting_minoverlap(self):
        temp_dir = tempfile.mkdtemp()
        try:
            mygprofiler = MagicMock()

            df = pd.DataFrame(columns=['name',
                                       'source',
                                       'native',
                                       'p_value',
                                       'description',
                                       'intersections',
                                       'precision',
                                       'recall',
                                       'term_size'],
                              data=[['name1',
                                     'source1',
                                     'native1',
                                     0.1,
                                     'desc1',
                                     ['hi'],
                                     0.5,
                                     0.7, 88],
                                    ['name2',
                                     'source2',
                                     'native2',
                                     0.5,
                                     'desc2',
                                     ['bye'],
                                     0.6,
                                     0.8, 99]])
            mygprofiler.profile = MagicMock(return_value=df)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c,')
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)
            theargs.minoverlap=0.6

            res = cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                          theargs,
                                                          gprofwrapper=
                                                          mygprofiler)
            self.assertEqual(None, res)

            mygprofiler.profile.assert_called_once_with(query=['a', 'b', 'c'],
                                                        domain_scope='known',
                                                        organism='hsapiens',
                                                        user_threshold=0.00000001,
                                                        no_evidences=False)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_with_no_result_cause_sources_in_exclude_list(self):
        temp_dir = tempfile.mkdtemp()
        try:
            mygprofiler = MagicMock()

            df = pd.DataFrame(columns=['name',
                                       'source',
                                       'native',
                                       'p_value',
                                       'description',
                                       'intersections',
                                       'precision',
                                       'recall',
                                       'term_size'],
                              data=[['name1',
                                     'TF',
                                     'native1',
                                     0.1,
                                     'desc1',
                                     ['hi'],
                                     0.5,
                                     0.7, 88],
                                    ['name2',
                                     'MIRNA',
                                     'native2',
                                     0.5,
                                     'desc2',
                                     ['bye'],
                                     0.6,
                                     0.8, 99],
                                    ['name2',
                                     'HP',
                                     'native2',
                                     0.5,
                                     'desc2',
                                     ['bye'],
                                     0.6,
                                     0.8, 99]
                                    ])
            mygprofiler.profile = MagicMock(return_value=df)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c,')
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)

            res = cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                          theargs,
                                                          gprofwrapper=
                                                          mygprofiler)
            self.assertEqual(None, res)

            mygprofiler.profile.assert_called_once_with(query=['a', 'b', 'c'],
                                                        domain_scope='known',
                                                        organism='hsapiens',
                                                        user_threshold=0.00000001,
                                                        no_evidences=False)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_with_valid_result(self):
        temp_dir = tempfile.mkdtemp()
        try:
            mygprofiler = MagicMock()

            df = pd.DataFrame(columns=['name',
                                       'source',
                                       'native',
                                       'p_value',
                                       'description',
                                       'intersections',
                                       'precision',
                                       'recall',
                                       'term_size'],
                              data=[['name1',
                                     'source1',
                                     'native1',
                                     0.1,
                                     'desc1',
                                     ['hi'],
                                     0.5,
                                     0.7, 88],
                                    ['name2',
                                     'source2',
                                     'native2',
                                     0.5,
                                     'desc2',
                                     ['bye'],
                                     0.6,
                                     0.8, 99]])
            mygprofiler.profile = MagicMock(return_value=df)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c,')
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)
            res = cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                          theargs,
                                                          gprofwrapper=
                                                          mygprofiler)
            self.assertEqual('name2', res['name'])
            self.assertEqual(['bye'], res['intersections'])
            self.assertEqual(99, res['term_size'])
            self.assertEqual(0.5, res['p_value'])
            self.assertEqual('source2', res['source'])
            self.assertEqual('native2', res['sourceTermId'])
            self.assertEqual(0.522, res['jaccard'])

            mygprofiler.profile.assert_called_once_with(query=['a', 'b', 'c'],
                                                        domain_scope='known',
                                                        organism='hsapiens',
                                                        user_threshold=0.00000001,
                                                        no_evidences=False)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_with_valid_no_intersections_in_result(self):
        temp_dir = tempfile.mkdtemp()
        try:
            mygprofiler = MagicMock()

            df = pd.DataFrame(columns=['name',
                                       'source',
                                       'native',
                                       'p_value',
                                       'description',
                                       'intersections',
                                       'precision',
                                       'recall',
                                       'term_size'],
                              data=[['name1',
                                     'source1',
                                     'native1',
                                     0.1,
                                     'desc1',
                                     ['hi'],
                                     0.5,
                                     0.7, 88],
                                    ['name2',
                                     'source2',
                                     'native2',
                                     0.5,
                                     'desc2',
                                     ['bye'],
                                     0.6,
                                     0.8, 99]])
            mygprofiler.profile = MagicMock(return_value=df)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c,')
            myargs = [tfile]
            theargs = cdgprofilergenestotermcmd._parse_arguments('desc',
                                                                 myargs)
            theargs.omit_intersections = True
            res = cdgprofilergenestotermcmd.run_gprofiler(tfile,
                                                          theargs,
                                                          gprofwrapper=
                                                          mygprofiler)
            self.assertEqual('name2', res['name'])
            self.assertEqual([], res['intersections'])
            self.assertEqual(99, res['term_size'])
            self.assertEqual(0.5, res['p_value'])
            self.assertEqual('source2', res['source'])
            self.assertEqual('native2', res['sourceTermId'])
            self.assertEqual(0.522, res['jaccard'])

            mygprofiler.profile.assert_called_once_with(query=['a', 'b', 'c'],
                                                        domain_scope='known',
                                                        organism='hsapiens',
                                                        user_threshold=0.00000001,
                                                        no_evidences=True)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_invalid_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = ['prog', tfile]
            res = cdgprofilergenestotermcmd.main(myargs)
            self.assertEqual(2, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = ['prog', tfile]
            res = cdgprofilergenestotermcmd.main(myargs)
            self.assertEqual(0, res)
        finally:
            shutil.rmtree(temp_dir)


if __name__ == '__main__':
    sys.exit(unittest.main())


cdgprofilergenestoterm/cdgprofilergenestoterm.py
#!/usr/bin/env python

import os
import sys
import argparse
import json
from gprofiler import GProfiler
import cdgprofilergenestoterm


def _parse_arguments(desc, args):
    """
    Parses command line arguments
    :param desc:
    :param args:
    :return:
    """
    help_fm = argparse.ArgumentDefaultsHelpFormatter
    parser = argparse.ArgumentParser(description=desc,
                                     formatter_class=help_fm)
    parser.add_argument('input',
                        help='comma delimited list of genes in file')
    parser.add_argument('--maxpval', type=float, default=0.00001,
                        help='Max p value')
    parser.add_argument('--organism', default='hsapiens',
                        help='Organism to use')
    return parser.parse_args(args)


def read_inputfile(inputfile):
    """

    :param inputfile:
    :return:
    """
    with open(inputfile, 'r') as f:
        return f.read()


def run_gprofiler(inputfile, theargs,
                  gprofwrapper=GProfiler(user_agent='cdgprofilergenestoterm/' +
                                         cdgprofilergenestoterm.__version__,
                                         return_dataframe=True)):
    """
    todo

    :param inputfile:
    :param theargs:
    :param gprofwrapper:
    :return:
    """
    genes = read_inputfile(inputfile)
    genes = genes.strip(',').strip('\n').split(',')
    if genes is None or (len(genes) == 1 and len(genes[0].strip()) == 0):
        sys.stderr.write('No genes found in input')
        return None
    df_result = gprofwrapper.profile(query=genes, domain_scope="known",
                                     organism=theargs.organism,
                                     user_threshold=theargs.maxpval,
                                     no_evidences=False)
    if df_result.shape[0] == 0:
        return None

    df_result['Jaccard'] = 1.0 / (1.0 / df_result['precision'] +
                                  1.0 / df_result['recall'] - 1)
    df_result.sort_values(['Jaccard', 'p_value'],
                          ascending=[False, True], inplace=True)
    df_result.reset_index(drop=True, inplace=True)
    theres = {'name': df_result['name'][0],
              'source': df_result['source'][0],
              'p_value': df_result['p_value'][0],
              'description': df_result['description'][0],
              'intersections': df_result['intersections'][0]}

    return theres


def main(args):
    """
    Main entry point for program

    :param args: command line arguments usually :py:const:`sys.argv`
    :return: 0 for success otherwise failure
    :rtype: int
    """
    desc = """
        Running gprofiler-official 1.0.0, with python3!

        Takes file with comma delimited list of genes as input and
        outputs matching term if any
    """

    theargs = _parse_arguments(desc, args[1:])

    try:
        inputfile = os.path.abspath(theargs.input)
        theres = run_gprofiler(inputfile, theargs)
        if theres is None:
            sys.stderr.write('No terms found\n')
        else:
            json.dump(theres, sys.stdout)
        sys.stdout.flush()
        return 0
    except Exception as e:
        sys.stderr.write('Caught exception: ' + str(e))
        return 2


if __name__ == '__main__':  # pragma: no cover
    sys.exit(main(sys.argv))


cdgprofilergenestoterm/cdgprofilergenestotermcmd.py
#!/usr/bin/env python

import os
import sys
import argparse
import json
from gprofiler import GProfiler
import cdgprofilergenestoterm

def _parse_arguments(desc, args):
    """
    Parses command line arguments
    :param desc:
    :param args:
    :return:
    """
    help_fm = argparse.ArgumentDefaultsHelpFormatter
    parser = argparse.ArgumentParser(description=desc,
                                     formatter_class=help_fm)
    parser.add_argument('input',
                        help='comma delimited list of genes in file')
    parser.add_argument('--maxpval', type=float, default=0.00000001,
                        help='Max p value')
    parser.add_argument('--minoverlap', default=0.05, type=float,
                        help='Minimum Jaccard to allow for hits')
    parser.add_argument('--omit_intersections', action='store_true',
                        help='If set, do NOT query for gene intersections')
    parser.add_argument('--excludesource', default='HP,MIRNA,TF',
                        help='Comma delimited list of sources to exclude')
    parser.add_argument('--maxgenelistsize', type=int,
                        default=500, help='Maximum number of genes that can'
                                          'be passed in via a query, '
                                          'exceeding this results in '
                                          'error')
    parser.add_argument('--precision', type=int, default=3,
                        help='Number of decimal places to round '
                             'jaccard')
    parser.add_argument('--organism', default='hsapiens',
                        help='Organism to use')
    return parser.parse_args(args)


def read_inputfile(inputfile):
    """

    :param inputfile:
    :return:
    """
    with open(inputfile, 'r') as f:
        return f.read()


def run_gprofiler(inputfile, theargs,
                  gprofwrapper=GProfiler(user_agent='cdgprofilergenestoterm/' +
                                         cdgprofilergenestoterm.__version__,
                                         return_dataframe=True)):
    """
    todo

    :param inputfile:
    :param theargs:
    :param gprofwrapper:
    :return:
    """
    genes = read_inputfile(inputfile)
    genes = genes.strip(',').strip('\n').split(',')
    genelist_size = len(genes)
    if genes is None or (genelist_size == 1 and len(genes[0].strip()) == 0):
        sys.stderr.write('No genes found in input')
        return None
    if genelist_size > theargs.maxgenelistsize:
        sys.stderr.write('Gene list size of ' +
                         str(genelist_size) +
                         ' exceeds max gene list size of ' +
                         str(theargs.maxgenelistsize))
        return None
    df_result = gprofwrapper.profile(query=genes, domain_scope="known",
                                     organism=theargs.organism,
                                     user_threshold=theargs.maxpval,
                                     no_evidences=theargs.omit_intersections)
    if df_result.shape[0] == 0:
        return None

    df_result['Jaccard'] = 1.0 / (1.0 / df_result['precision'] +
                                  1.0 / df_result['recall'] - 1)

    # filter out any rows where min overlap is not met
    df_result.drop(df_result[df_result.Jaccard < theargs.minoverlap].index,
                   inplace=True)

    # filter out any rows in source exclude list
    if theargs.excludesource is not None:
        for a_source in theargs.excludesource.split(','):
            df_result.drop(df_result[df_result.source == a_source].index,
                           inplace=True)

    if df_result.shape[0] == 0:
        return None

    # sort by Jaccard and fallback to p_value
    df_result.sort_values(['Jaccard', 'p_value'],
                          ascending=[False, True], inplace=True)

    df_result.reset_index(drop=True, inplace=True)

    df_result = df_result[:1]
    df_result = df_result.round({'Jaccard': theargs.precision})
    theres = {'name': df_result['name'][0],
              'source': df_result['source'][0],
              'sourceTermId': df_result['native'][0],
              'p_value': df_result['p_value'][0],
              'jaccard': df_result['Jaccard'][0],
              'description': df_result['description'][0],
              'term_size': int(df_result['term_size'][0])}

    if theargs.omit_intersections is True:
        theres['intersections'] = []
    else:
        theres['intersections'] = df_result['intersections'][0]
    return theres


def main(args):
    """
    Main entry point for program

    :param args: command line arguments usually :py:const:`sys.argv`
    :return: 0 for success otherwise failure
    :rtype: int
    """
    desc = """
        Using gprofiler-official 1.0.0, Python module, this
        program takes a file with comma delimited list of genes 
        as input and outputs best matching term in JSON format to
        standard out. Any log messages and errors are output to
        standard error.
        
        Return 0 upon success otherwise error.
        
        Format of JSON output:
        
        {
         "name": "<TERM NAME>",
         "source": "<IS THE NAME FOR THE DATASOURCE>",
         "sourceTermId": "<IS THE ID FOR THE ENRICHED TERM/FUNCTIONAL CATEGORY IN ITS NATIVE NAMESPACE>",
         "p_value": <PVALUE>,
         "jaccard": <JACCARD VALUE>,
         "description": "<DESCRIPTION, IF ANY, FOR TERM>",
         "term_size": <NUMBER OF GENES ASSOCIATED WITH TERM>,
         "intersections": ["<LIST OF GENES USED TO GET TERM>"]
        }

    """

    theargs = _parse_arguments(desc, args[1:])

    try:
        inputfile = os.path.abspath(theargs.input)
        theres = run_gprofiler(inputfile, theargs)
        if theres is None:
            sys.stderr.write('No terms found\n')
        else:
            json.dump(theres, sys.stdout)
        sys.stdout.flush()
        return 0
    except Exception as e:
        sys.stderr.write('Caught exception: ' + str(e))
        return 2


if __name__ == '__main__':  # pragma: no cover
    sys.exit(main(sys.argv))
